firefox-ios/Client/Application/SceneDelegate.swift

```diff
@@ -148,21 +190,57 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
 
         // Check if our connection options include a user response to a push
         // notification that is for Sent Tabs. If so, route the related tab URLs.
+        if let sentTabsUserInfo = sentTabsUserInfoFromConnectionOptions(connectionOptions) {
+            handleConnectionOptionsSentTabs(sentTabsUserInfo)
+        }
+    }
+
+    private func sentTabsUserInfoFromConnectionOptions(_ options: UIScene.ConnectionOptions) -> [[String: Any]]? {
         let sentTabsKey = NotificationSentTabs.sentTabsKey
-        if let notification = connectionOptions.notificationResponse?.notification,
+        if let notification = options.notificationResponse?.notification,
            let userInfo = notification.request.content.userInfo[sentTabsKey] as? [[String: Any]] {
-            handleConnectionOptionsSentTabs(userInfo)
+            return userInfo
+        }
+        return nil
+    }
+
+    private func sentTabsURLs(for userInfo: [[String: Any]]?) -> [URL] {
+        guard let userInfo else { return [] }
+        return userInfo.compactMap {
+            guard let urlString = $0["url"] as? String, let url = URL(string: urlString) else { return nil }
+            return url
         }
     }
 
     private func handleConnectionOptionsSentTabs(_ userInfo: [[String: Any]]) {
         // For Sent Tab data structure, see also:
         // NotificationService.displayNewSentTabNotification()
-        for tab in userInfo {
-            guard let urlString = tab["url"] as? String,
-                  let url = URL(string: urlString),
-                  let route = routeBuilder.makeRoute(url: url) else { continue }
+        let urls = sentTabsURLs(for: userInfo)
+        urls.forEach {
+            guard let route = routeBuilder.makeRoute(url: $0) else { return }
             sceneCoordinator?.findAndHandle(route: route)
         }
     }
+
+    private func logEventForExpectedDeeplink(url: URL, windowUUID: WindowUUID) {
+        guard let urlScanner = URLScanner(url: url) else { return }
+        let targetURL: URL
+        if urlScanner.isOurScheme {
+            guard let parsedURL = urlScanner.fullURLQueryItem()?.asURL else { return }
+            targetURL = parsedURL
+        } else {
+            targetURL = url
+        }
+
+        let expectedEvent: AppEvent = .selectTab(targetURL, windowUUID)
+        DispatchQueue.main.asyncAfter(deadline: .now() + 4.0) { [weak self] in
+            guard let self else { return }
+            if AppEventQueue.hasSignalled(expectedEvent) {
+                logger.log("Deeplink tab check: success.", level: .debug, category: .deeplinks)
+            } else {
+                // Indicates a Sent Tab or deeplink that failed to open. Log to Sentry. [FXIOS-8374]
+                logger.log("Deeplink tab check: failed.", level: .fatal, category: .deeplinks)
```

Comment:
This has me a little nervous. If we uplift this to 123 we can run it in beta for at least a week and then see if it's very noisy, that will give us time to pull it out before it goes to prod if needed.